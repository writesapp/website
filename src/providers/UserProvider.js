import React, { useState, useEffect } from 'react';
import { auth } from '../firebaseInit';

export const UserContext = React.createContext({ user: null });

export function UserProvider({ children }) {
  const [user, setUser] = useState(null);

  useEffect(() => {
    auth.onAuthStateChanged(user => setUser(user))
  }, []);

  return <UserContext.Provider value={{ user }}>{children}</UserContext.Provider>
}
