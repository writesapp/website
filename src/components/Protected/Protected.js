import { useEffect, useContext } from 'react';
import { UserContext } from '../../providers/UserProvider';
import { navigate } from '@reach/router';

export default function Protected({ view }) {
  const { user } = useContext(UserContext);

  useEffect(() => {
    if (!user) navigate('/login', { replace: true });
  });

  return user ? view : null;
}