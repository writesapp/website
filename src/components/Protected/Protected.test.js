import React from "react";
import { render, cleanup } from "@testing-library/react";

import Protected from "./Protected";
import { UserContext } from "../../providers/UserProvider";

function renderWithContext(fragment, user) {
  return render(<UserContext.Provider value={{ user }}>{fragment}</UserContext.Provider>);
}

function Foo() {
  return <div data-testid="bar-view">bar</div>;
}

const mockUser = { email: "a@siekierski.ml" };

afterEach(() => {
  cleanup();
});

it("renders without crashing, when passed user object", () => {
  renderWithContext(<Protected view={Foo} />, mockUser);
});

it("renders without crashing, when didn't pass user object", () => {
  renderWithContext(<Protected view={Foo} />);
});

it("renders the view properly", () => {
  const { getByTestId } = renderWithContext(<Protected view={Foo} />, mockUser);

  expect(getByTestId("bar-view")).toBeInTheDocument();
  expect(getByTestId("bar-view").textContent).toBe("bar");
});

it("renders the view without changing anything", () => {
  const { container } = renderWithContext(<Foo />, mockUser);

  expect(container).toMatchSnapshot();
});
