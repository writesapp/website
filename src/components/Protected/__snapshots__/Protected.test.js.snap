// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the view without changing anything 1`] = `
<div>
  <div
    data-testid="bar-view"
  >
    bar
  </div>
</div>
`;
